### PR TITLE
Model refactor

### PR DESCRIPTION
**Pull request by jindrahelcl**
Read the commit messages.

ModelPart is now a syntactic sugar class for these three abstract classes:
`Parameterized` - manages variables, name scopes, loading and saving
`Feedable` - has feed_dict, batch_size, and train_mode (in the future, may also contain some logic for creating placeholders or handling data in tf.dataset)
`GenericModelPart` - the get_dependencies function is now here and all the classes that handle something in the graph before runners and/or trainers are descendants of this class

as a bonus, this PR fixes the dependenies gradient blocking objects and provides a way of how to do 
it e.g. for adversarial gradient layers introduced in #771

### CODE REVIEW

File: neuralmonkey/model/model_part.py [hunk 1]
```diff
@@ -1,147 +1,84 @@
 """Basic functionality of all model parts."""
-
 from abc import ABCMeta
-from contextlib import contextmanager
-from typing import Any, Callable, Dict, List, Set, Tuple
+from typing import MutableSet, Set, List, Tuple
+
+# pylint: disable=unused-import
+# TODO feed dict and initializer specs are imported for convenience
+# TODO this is because the codebase import these things from this module
```
Comment (jindrahelcl) [inline neuralmonkey/model/model_part.py]: Todle by se ještě dalo dodělat...

**Comment by jindrahelcl** [comment]
Zdokumentováno...

File: neuralmonkey/model/parameterized.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        self._save_checkpoint = save_checkpoint
+        self._load_checkpoint = load_checkpoint
+
+        self._saver = None  # type: tf.train.Saver
+        self._reuse = reuse is not None
+
+        if reuse is not None:
+            # pylint: disable=unidiomatic-typecheck
+            # Here we need an exact match of types
+            if type(self) != type(reuse):
```
Comment (varisd) [inline neuralmonkey/model/parameterized.py]: Tyhle radky asi muzem smazat, viz #775

**Comment by jindrahelcl** [comment]
This is ready for review.

File: neuralmonkey/model/model_part.py [hunk 3]
```diff
@@ -1,147 +1,114 @@
...
+    """
+
+    @property
+    def list_dependencies(self) -> List[str]:
+        """Return a list of attributes regarded as lists of dependents."""
+        return ["attentions", "encoders"]
+
+    @property
+    def singleton_dependencies(self) -> List[str]:
+        """Return a list of attributes regarded as dependents."""
```
Comment (jlibovicky) [inline neuralmonkey/model/model_part.py]: List of attribute names

File: neuralmonkey/model/model_part.py [hunk 4]
```diff
@@ -1,147 +1,114 @@
...
 
-import tensorflow as tf
+    Apart from the `get_dependencies` method, this class also provides two
+    properties: `list_dependencies` and `singleton_dependencies` which store
+    the names of the Python class attributes that are regarded as potential
+    dependents of the `GenericModelPart` object. These dependents are
+    automatically checked for type and when they are instances of the
+    `GenericModelPart` class, results of their `get_dependencies` are united
+    and returned as dependencies of the current object.
+    """
```
Comment (jlibovicky) [inline neuralmonkey/model/model_part.py]: Tenhleta konstrukce se mi moc nelíbí. IMHO by stačilo mít property, která se jmenuje dependencies, a když by ten atribut byl list nebo obecně iterable, tak by se proiteroval, tohle IHMO přidává cognitive load čtenářům těch model partů.

Kacíčská myšlenka: Dovedeme si představit situaci, kde je něco atribut třídy, ale z nějakého důvodu to není dependency. Pokud ne, tak bych prostě v `get_dependencies` iteroval přes všechny atributy třídy a zachoval bych se podle jejich typu.

File: neuralmonkey/model/model_part.py [hunk 5]
```diff
@@ -1,147 +1,114 @@
...
+
+    @property
+    def list_dependencies(self) -> List[str]:
+        """Return a list of attributes regarded as lists of dependents."""
+        return ["attentions", "encoders"]
+
+    @property
+    def singleton_dependencies(self) -> List[str]:
```
Comment (jlibovicky) [inline neuralmonkey/model/model_part.py]: Název singleton IMHO evokuje ten Javařskej návrhovej vzor, tak by to chtělo lepší název, i kdyby to zůstalo.

File: neuralmonkey/model/sequence_split.py [hunk 6]
```diff
@@ -71,14 +72,12 @@ def temporal_mask(self) -> tf.Tensor:
         return tf.squeeze(
             split_by_factor(double_mask, self.batch_size, self.factor), axis=2)
 
-    def feed_dict(self, dataset: Dataset, train: bool) -> FeedDict:
+    def feed_dict(self, dataset: Dataset, train: bool = True) -> FeedDict:
```
Comment (jlibovicky) [inline neuralmonkey/model/sequence_split.py]: Vážně je správě ta mezera okolo `=`?

Comment (jindrahelcl) [reply neuralmonkey/model/sequence_split.py]: ano.

File: neuralmonkey/model/model_part.py (hunk 5 above)
Comment (jindrahelcl) [reply neuralmonkey/model/model_part.py]: vyhodil jsem to a zůstalo jen property dependencies

File: neuralmonkey/model/model_part.py (hunk 4 above)
Comment (jindrahelcl) [reply neuralmonkey/model/model_part.py]: ad konstrukce: hotovo, je tam jen dependencies.
ad kacíři: Radši bych tam ty dependencies měl takhle napůl explicitně a třeba se na to detailně podíval v další refactor iteraci..

File: neuralmonkey/model/model_part.py (hunk 3 above)
Comment (jindrahelcl) [reply neuralmonkey/model/model_part.py]: hotovo
